### ISSUE TITLE
Possible Keras Implementation doesn't seem to work

### ISSUE THREAD

**Issue by pgadosey** [joe-siyuan-qiao/WeightStandardization#12]
Hi, thank you very much for your work, I attempted to reproduce this in keras using the following code but doesn't seem to improve my model. Any suggestions on what i might be doing wrong?

    def ws_function(input_tensor):
    kernel_mean = Lambda(lambda x:tf.reduce_mean(input_tensor, axis=[0, 1, 2], keep_dims=True, name='kernel_mean'))(input_tensor)
    kernel =  Lambda(lambda x: x[0]-x[1])([input_tensor, kernel_mean])
    #kernel = Lambda(lambda x: x[0]-x[1])([input_tensor,kernel_mean])
    kernel_std = Lambda(lambda x: tf.keras.backend.std(kernel, axis=[0, 1, 2], keepdims=True))(kernel)
    #kernel = kernel / (kernel_std + 1e-5) 
    kernel =  Lambda(lambda x: x[0]/x[1])([kernel, (kernel_std + 1e-5)])
    #print(type(kernel))
    return kernel

    def ws_conv2d(input_tensor, n_filters,kernel_size=3,batchnorm=True):
    kernel= Lambda(lambda x: ws_function(input_tensor))(input_tensor)
    #kernel = ws_function(input_tensor)
    # first layer
    x = Conv2D(filters = n_filters, kernel_size = (kernel_size, kernel_size),\
              kernel_initializer = 'he_normal', padding = 'same')(kernel)
    x = Activation('relu')(x)
    
    # second layer
    x = Lambda(lambda x: ws_function(x))(x)
    #kernel = ws_function(x)
    x = Conv2D(filters = n_filters, kernel_size = (kernel_size, kernel_size),\
              kernel_initializer = 'he_normal', padding = 'same')(x)
    if batchnorm:
        x = conv1 = GroupNormalization(group=32)(x)
    x = Activation('relu')(x)
    #print(type(x))
    return x

**Comment by csrhddlam** [comment]
Hi, thanks for your interest. I'm not familiar with Keras, but it seems you are normalizing the input_tensor, i.e. doing a batch normalization, rather than the convolutional kernel. Correct me if I'm wrong. See Figure 2 for the comparison between activation normalization methods and WS.

**Comment by zihaozhihao** [comment]
Like @csrhddlam said, you are normalizing the input instead of the kernel. A simple workaround is to overwrite the `call()` method of `Conv2D`, then simply change your `Conv2D` to `Conv2D_WS`. The following is the code I used. 
```python
class Conv2D_WS(Conv2D):
    def __init__(self, filters,
                 kernel_size,
                 strides=(1, 1),
                 padding='valid',
                 data_format=None,
                 dilation_rate=(1, 1),
                 activation=None,
                 use_bias=True,
                 kernel_initializer='glorot_uniform',
                 bias_initializer='zeros',
                 kernel_regularizer=None,
                 bias_regularizer=None,
                 activity_regularizer=None,
                 kernel_constraint=None,
                 bias_constraint=None,
                 **kwargs):
        super(Conv2D_WS, self).__init__(filters=filters,
                 kernel_size=kernel_size,
                 strides=strides,
                 padding=padding,
                 data_format=data_format,
                 dilation_rate=dilation_rate,
                 activation=activation,
                 use_bias=use_bias,
                 kernel_initializer=kernel_initializer,
                 bias_initializer=bias_initializer,
                 kernel_regularizer=kernel_regularizer,
                 bias_regularizer=bias_regularizer,
                 activity_regularizer=activity_regularizer,
                 kernel_constraint=kernel_constraint,
                 bias_constraint=bias_constraint,
                 **kwargs)

    def call(self, inputs):
        mean = K.mean(self.kernel, axis=[0,1,2], keepdims=True)
        self.kernel = self.kernel - mean
        std = K.sqrt(K.var(self.kernel, axis=[0,1,2], keepdims=True) + 1e-12)
#         std = K.std(self.kernel, axis=[0,1,2], keepdims=True) # loss will be nan after several epochs. 
        self.kernel = self.kernel / (std + 1e-5)

        outputs = K.conv2d(
            inputs,
            self.kernel,
            strides=self.strides,
            padding=self.padding,
            data_format=self.data_format,
            dilation_rate=self.dilation_rate)
        
        if self.use_bias:
            outputs = K.bias_add(
                outputs,
                self.bias,
                data_format=self.data_format)

        if self.activation is not None:
            return self.activation(outputs)
        return outputs
```
@csrhddlam Sometimes, when I train the model with small batch size, such as 4, the model cannot be converged. Have you ever had the same issue before?

**Comment by csrhddlam** [comment]
By "cannot converge", did you mean "The loss does not decrease at all after initialization"? We haven't observed such case on models and tasks that we experimented on. Are you using BN or GN? Does the model converge without WS?

**Comment by zihaozhihao** [comment]
Sometimes, after training several epochs, the loss will be nan. I'm using BN, and model converge without WS.

**Comment by joe-siyuan-qiao** [comment]
@zihaozhihao This issue was mentioned #1. Please take a look and see if the solution helps. Thanks.

**Comment by zihaozhihao** [comment]
@joe-siyuan-qiao The solution does help for now. I have updated my keras solution above. Thanks a lot!

**Comment by pgadosey** [comment]
@zihaozhihao Apologies for the late reply. I just decided to revisit this repo and so your reply. Thanks
